dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/EventDateValidationHook.java

```diff
@@ -0,0 +1,205 @@
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import org.hisp.dhis.event.EventStatus;
+import org.hisp.dhis.period.Period;
+import org.hisp.dhis.period.PeriodType;
+import org.hisp.dhis.program.Program;
+import org.hisp.dhis.program.ProgramStageInstance;
+import org.hisp.dhis.tracker.TrackerImportStrategy;
+import org.hisp.dhis.tracker.domain.Event;
+import org.hisp.dhis.tracker.report.TrackerErrorCode;
+import org.hisp.dhis.tracker.report.ValidationErrorReporter;
+import org.hisp.dhis.tracker.validation.TrackerImportValidationContext;
+import org.hisp.dhis.user.User;
+import org.hisp.dhis.util.DateUtils;
+import org.springframework.stereotype.Component;
+
+import java.util.Date;
+
+import static com.google.api.client.util.Preconditions.checkNotNull;
+import static org.hisp.dhis.tracker.report.ValidationErrorReporter.newReport;
+
+/**
+ * @author Morten Svanæs <msvanaes@dhis2.org>
+ */
+@Component
+public class EventDateValidationHook
+    extends AbstractTrackerDtoValidationHook
+{
+    public EventDateValidationHook()
+    {
+        super( Event.class, TrackerImportStrategy.CREATE_AND_UPDATE );
+    }
+
+    @Override
+    public void validateEvent( ValidationErrorReporter reporter, Event event )
+    {
+        TrackerImportValidationContext context = reporter.getValidationContext();
+
+        if ( EventStatus.ACTIVE == event.getStatus() && event.getOccurredAt() == null )
+        {
+            reporter.addError( newReport( TrackerErrorCode.E1031 )
+                .addArg( event ) );
+            return;
+        }
+
+        ProgramStageInstance programStageInstance = context.getProgramStageInstance( event.getEvent() );
+        Program program = context.getProgram( event.getProgram() );
+
+        validateDateFormat( reporter, event );
+        validateExpiryDays( reporter, event, program, programStageInstance, context.getBundle().getUser() );
+        validatePeriodType( reporter, event, program, programStageInstance );
+    }
+
+    private void validateExpiryDays( ValidationErrorReporter reporter, Event event, Program program,
+        ProgramStageInstance programStageInstance, User actingUser )
+    {
+        checkNotNull( actingUser, TrackerImporterAssertErrors.USER_CANT_BE_NULL );
+        checkNotNull( event, TrackerImporterAssertErrors.EVENT_CANT_BE_NULL );
+        checkNotNull( program, TrackerImporterAssertErrors.PROGRAM_CANT_BE_NULL );
+
+        if ( (program.getCompleteEventsExpiryDays() > 0 && EventStatus.COMPLETED == event.getStatus())
+            || (programStageInstance != null && EventStatus.COMPLETED == programStageInstance.getStatus()) )
```

Comment:
What do you mean not present? getCompleteEventsExpiryDays is a int, it can never be null. I'm not sure what you mean here, I think we need to discuss this in a meeting. Adding to meeting document.